### PR TITLE
feat(wdio-junit-reporter): allow to override className format

### PR DESCRIPTION
**Pull request by facusantillo**
## Proposed changes
Issue: #7645

Allow overriding name attribute on testsuite tags with a function like it was proposed in the related issue.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/main/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [ ] I have added proper type definitions for new commands (if appropriate)

## Further comments

[//]: # (If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...)

### CODE REVIEW

**Comment by facusantillo** [comment]
@christian-bromann I didn't have my GH correctly configured. I've signed now to EasyCLA. Is there any additional steps that I must do? 

Also, I created the snapshots and tested them but am not sure codewise if everything is ok. 

Please let me know :)

File: packages/wdio-junit-reporter/src/types.ts [hunk 1]
```diff
@@ -49,6 +55,15 @@ export interface JUnitReporterOptions extends Reporters.Options {
...
     classNameFormat?: (options: ClassNameFormatOptions) => string
+    /**
+     * Give the ability to override the generated name of a test suite. suiteNameFormat RegExp would be applied to the suite name for normalization.
+     *
+     * @default
+     * testSuiteNameFormatOptions: function (options) {
+     *     return ``${options.suite.title}``
+     * }
+     */
+    testSuiteName?: (options: TestSuiteNameFormatOptions) => string
```
Comment (christian-bromann) [inline packages/wdio-junit-reporter/src/types.ts:66]:
```suggestion
    suiteNameFormat?: (options: TestSuiteNameFormatOptions) => string
```

File: packages/wdio-junit-reporter/src/types.ts [hunk 2]
```diff
@@ -49,6 +55,15 @@ export interface JUnitReporterOptions extends Reporters.Options {
      * }
      */
     classNameFormat?: (options: ClassNameFormatOptions) => string
+    /**
+     * Give the ability to override the generated name of a test suite. suiteNameFormat RegExp would be applied to the suite name for normalization.
+     *
+     * @default
+     * testSuiteNameFormatOptions: function (options) {
```
Comment (christian-bromann) [inline packages/wdio-junit-reporter/src/types.ts:62]:
```suggestion
     * suiteNameFormat: function (options) {
```

File: packages/wdio-junit-reporter/src/index.ts [hunk 3]
```diff
@@ -149,9 +149,9 @@ class JunitReporter extends WDIOReporter {
         specFileName: string,
         suite: SuiteStats
     ) {
-        const suiteName = this._prepareName(suite.title)
+        //const suiteName =
         const filePath = specFileName.replace(process.cwd(), '.')
-
+        const suiteName = this.options.testSuiteName ? this._prepareName(this.options.testSuiteName({ name: this.options.testSuiteName.name, suite })) : this._prepareName(suite.title)
```
Comment (christian-bromann) [inline packages/wdio-junit-reporter/src/index.ts:154]: What is `this.options.testSuiteName` here? A function or an object with the property `name`?

**Review by christian-bromann — CHANGES_REQUESTED** [review]
Some comments. Please try to rebase and rename the commit so they will be signed with your correct identity. See also https://git-scm.com/book/en/v2/Getting-Started-First-Time-Git-Setup

File: packages/wdio-junit-reporter/src/types.ts (hunk 1 above)
Comment (facusantillo) [reply packages/wdio-junit-reporter/src/types.ts:66]: suiteNameFormat is already in use by the regex format, that's why I choose this name

File: packages/wdio-junit-reporter/src/types.ts (hunk 2 above)
Comment (facusantillo) [reply packages/wdio-junit-reporter/src/types.ts:62]: testSuiteNameFormatOptions is already in use

File: packages/wdio-junit-reporter/src/index.ts (hunk 3 above)
Comment (facusantillo) [reply packages/wdio-junit-reporter/src/index.ts:154]: Is the function defined before

**Comment by facusantillo** [comment]
I rebased the branch and did what's recommended in the documentation. Although EasyCLA is still not validating my identity

File: packages/wdio-junit-reporter/src/index.ts [hunk 4]
```diff
@@ -149,9 +149,9 @@ class JunitReporter extends WDIOReporter {
         specFileName: string,
         suite: SuiteStats
     ) {
-        const suiteName = this._prepareName(suite.title)
+        //const suiteName =
```
Comment (christian-bromann) [inline packages/wdio-junit-reporter/src/index.ts:152]: Do we need this?
```suggestion
```

File: packages/wdio-junit-reporter/src/index.ts [hunk 5]
```diff
@@ -69,7 +69,7 @@ class JunitReporter extends WDIOReporter {
         specFileName: string,
         suite: SuiteStats
     ) {
-        const featureName = this._prepareName(suite.title)
+        const featureName = this.options.testSuiteName ? this._prepareName(this.options.testSuiteName({ name: this.options.testSuiteName.name, activeFeatureName: this._activeFeatureName })): this._prepareName(suite.title)
```
Comment (christian-bromann) [inline packages/wdio-junit-reporter/src/index.ts:72]: If someone provides a `testSuiteName` function we should expect that the name is valid - no need to call `_prepareName` and have some weird side effects if also `suiteNameFormat` is provided.

```suggestion
        const featureName = typeof this.options.testSuiteName === 'function'
            ? this.options.testSuiteName({
                name: this.options.testSuiteName.name,
                activeFeatureName: this._activeFeatureName
            })
            : this._prepareName(suite.title)
```

File: packages/wdio-junit-reporter/src/index.ts (hunk 3 above)
Comment (christian-bromann) [reply packages/wdio-junit-reporter/src/index.ts:154]: Should we stay consistent and call the second argument `activeSuiteName`?

```suggestion
        const suiteName = this.options.testSuiteName
            ? this.options.testSuiteName({
                name: this.options.testSuiteName.name,
                suite
             })
             : this._prepareName(suite.title)
```

File: packages/wdio-junit-reporter/src/types.ts [hunk 6]
```diff
@@ -16,6 +16,12 @@ interface ClassNameFormatOptions {
     suite?: SuiteStats,
 }
 
+interface TestSuiteNameFormatOptions {
+    name?: string,
+    activeFeatureName?: any
+    suite?: SuiteStats
```
Comment (christian-bromann) [inline packages/wdio-junit-reporter/src/types.ts:22]: I suggest we find some more consistency here. Do we need two parameters? Can we maybe settle on one name for it?

File: packages/wdio-junit-reporter/src/types.ts (hunk 1 above)
Comment (christian-bromann) [reply packages/wdio-junit-reporter/src/types.ts:66]: How about we allow it to be `RegExp | Function` so we have one unified parameter for formatting the suite name?

**Review by christian-bromann — CHANGES_REQUESTED** [review]

File: packages/wdio-junit-reporter/src/types.ts [hunk 7]
```diff
@@ -13,12 +15,11 @@ interface ClassNameFormatOptions {
     /**
      * Context of the current suite
      */
-    suite?: SuiteStats,
+    suite?: SuiteStats
```
Comment (christian-bromann) [inline packages/wdio-junit-reporter/src/types.ts:18]: `suite` seems to be always given, so let's improve this.

```suggestion
    suite: SuiteStats
```

File: packages/wdio-junit-reporter/src/index.ts [hunk 8]
```diff
@@ -69,7 +69,8 @@ class JunitReporter extends WDIOReporter {
         specFileName: string,
         suite: SuiteStats
     ) {
-        const featureName = this.options.testSuiteName ? this._prepareName(this.options.testSuiteName({ name: this.options.testSuiteName.name, activeFeatureName: this._activeFeatureName })): this._prepareName(suite.title)
+        const featureName = this.options.suiteNameFormat instanceof RegExp || !this.options.suiteNameFormat  ? this._prepareName(suite.title) : this.options.suiteNameFormat({ name: this.options.suiteNameFormat.name, suite })
```
Comment (christian-bromann) [inline packages/wdio-junit-reporter/src/index.ts:72]: minor formatting 

```suggestion
        const featureName = !this.options.suiteNameFormat || this.options.suiteNameFormat instanceof RegExp
            ? this._prepareName(suite.title)
            : this.options.suiteNameFormat({ name: this.options.suiteNameFormat.name, suite })
```

File: packages/wdio-junit-reporter/src/index.ts [hunk 9]
```diff
@@ -149,9 +150,9 @@ class JunitReporter extends WDIOReporter {
         specFileName: string,
         suite: SuiteStats
     ) {
-        //const suiteName =
+
         const filePath = specFileName.replace(process.cwd(), '.')
-        const suiteName = this.options.testSuiteName ? this._prepareName(this.options.testSuiteName({ name: this.options.testSuiteName.name, suite })) : this._prepareName(suite.title)
+        const suiteName = this.options.suiteNameFormat instanceof RegExp || !this.options.suiteNameFormat  ? this._prepareName(suite.title) : this.options.suiteNameFormat({ name: this.options.suiteNameFormat.name, suite })
```
Comment (christian-bromann) [inline packages/wdio-junit-reporter/src/index.ts:155]: same
```suggestion
        const suiteName = !this.options.suiteNameFormat || this.options.suiteNameFormat instanceof RegExp
            ? this._prepareName(suite.title)
            : this.options.suiteNameFormat({ name: this.options.suiteNameFormat.name, suite })
```

File: packages/wdio-junit-reporter/src/types.ts [hunk 10]
```diff
@@ -1,6 +1,8 @@
 import { Reporters } from '@wdio/types'
 import { SuiteStats } from '@wdio/reporter'
 
+type testSuiteName = (options: TestSuiteNameFormatOptions) => string
```
Comment (christian-bromann) [inline packages/wdio-junit-reporter/src/types.ts:4]:
```suggestion
type TestSuiteNameFunction = (options: TestSuiteNameFormatOptions) => string
```

File: packages/wdio-junit-reporter/src/types.ts [hunk 11]
```diff
@@ -41,11 +42,19 @@ export interface JUnitReporterOptions extends Reporters.Options {
...
+     * @example regex
+     * /[^a-z0-9]+/
+     *
+     * @example suiteName
+     * suiteNameFormat: function (options) {
+     *     return `${options.suite.title}`
+     * }
      */
-    suiteNameFormat?: RegExp
+    suiteNameFormat?: RegExp | testSuiteName
```
Comment (christian-bromann) [inline packages/wdio-junit-reporter/src/types.ts:57]:
```suggestion
    suiteNameFormat?: RegExp | TestSuiteNameFunction
```

**Review by christian-bromann — COMMENTED** [review]
some more comments

**Review by christian-bromann — APPROVED** [review]
Thanks a lot 👍

**Comment by christian-bromann** [comment]
@facusantillo it seems that the EasyCLI bot still can't detect your identity. I recommend to squash all commits into one and resign them. If this doesn't work, maybe start from a fresh branch and copy all changes over. I can't merge the commit without having the CLA signed.

**Comment by facusantillo** [comment]
Seems that EasyCLA it's hang because of the previous issue, will close this PR and start a new one

**Pull request closed without merging**